### PR TITLE
Moved to and from class methods from Rotation to Quaternion

### PR DESCRIPTION
**Pull request by anderscmathisen**
#### Description of the change

Because of discussions in #401,  #345 and #373, it seemed like there is a need to move the `to_` and `from_` methods of `Rotation` to the `Quaternion` class. And so I have copy pasted them over.

I have changed little to nothing about the methods themselves, other than changing names of variables and updated docstrings. 

I did not move the `random_vonmises()` method as that would also require the `angle_with` method of `Rotation` class to be moved as well, and I was unsure whether that should be done or not. 

I included an override of `from_euler` in `Rotation` to keep the proper vs improper rotation notion as it was originally.

#### Progress of the PR
- [x] [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- [x] Unit tests with pytest for all lines
- [x] Clean code style by [running black via pre-commit](https://orix.readthedocs.io/en/latest/contributing.html#code-style)

#### Minimal example of the bug fix or new feature
```python
>>> from orix.quaternion import Quaternion, Rotation
>>> import numpy as np
>>> euler = np.random.rand(10, 3)
>>> rot = Rotation.from_euler(euler)
>>> quat = Quaternion.from_euler(euler)
```

#### For reviewers

- [ ] The PR title is short, concise, and will make sense 1 year later.
- [ ] New functions are imported in corresponding `__init__.py`.
- [ ] New features, API changes, and deprecations are mentioned in the unreleased
      section in `CHANGELOG.rst`.
- [ ] Contributor(s) are listed correctly in `__credits__` in `orix/__init__.py` and in
      `.zenodo.json`.

### CODE REVIEW

File: orix/quaternion/quaternion.py [hunk 1]
```diff
@@ -498,3 +502,311 @@ def _outer_dask(
...
+        om[..., 1, 1] = qq + 2 * cc
+        om[..., 1, 2] = 2 * (cd - ab)
+        om[..., 2, 0] = 2 * (bd - ac)
+        om[..., 2, 1] = 2 * (cd + ab)
+        om[..., 2, 2] = qq + 2 * dd
+
+        return om
+
+    @classmethod
+    def from_matrix(cls, matrix: np.ndarray) -> Quaternion:
```
Comment (hakonanes) [inline orix/quaternion/quaternion.py:743]:
```suggestion
    def from_matrix(cls, matrix: Union[np.ndarray, tuple, list]) -> Quaternion:
```

File: orix/quaternion/quaternion.py [hunk 2]
```diff
@@ -498,3 +502,311 @@ def _outer_dask(
...
+        """Return quaternions from the orientation matrices
+        :cite:`rowenhorst2015consistent`.
+
+        Parameters
```
Comment (hakonanes) [inline orix/quaternion/quaternion.py:747]: We need a `Returns`!

File: orix/quaternion/quaternion.py [hunk 3]
```diff
@@ -498,3 +502,311 @@ def _outer_dask(
...
+        >>> np.allclose(rot.data, [1, 0, 0, 0])
+        True
+        >>> rot = Quaternion.from_matrix(np.diag([1, -1, -1]))
+        >>> np.allclose(rot.data, [0, 1, 0, 0])
+        True
+        """
+        om = np.asarray(matrix)
+        # Assuming (3, 3) as last two dims
+        n = (1,) if om.ndim == 2 else om.shape[:-2]
+        quat = np.zeros(n + (4,))
```
Comment (hakonanes) [inline orix/quaternion/quaternion.py:765]: We can replace this check with

```suggestion
        om = np.atleast_3d(matrix)
        quat = np.zeros(om.shape[:-2] + (4,))
```

File: orix/quaternion/rotation.py [hunk 4]
```diff
@@ -385,133 +380,26 @@ def dot_outer(self, other: Rotation) -> np.ndarray:
...
-        return rot
-
-    @classmethod
-    def from_axes_angles(
-        cls,
-        axes: Union[np.ndarray, Vector3d, tuple, list],
-        angles: Union[np.ndarray, tuple, list],
-    ) -> Rotation:
-        """Create rotation(s) from axis-angle pair(s).
+    def identity(cls, shape: tuple = (1,)) -> Rotation:
```
Comment (hakonanes) [inline orix/quaternion/rotation.py:383]: Would be good to allow `shape` to be an integer as in `Quaternion.random()` (see above).

File: orix/quaternion/rotation.py [hunk 5]
```diff
@@ -385,133 +380,26 @@ def dot_outer(self, other: Rotation) -> np.ndarray:
...
+        """Create identity rotations.
 
         Parameters
         ----------
-        axes
-            The axis of rotation.
-        angles
-            The angle of rotation, in radians.
+        shape
+            The shape out of which to construct identity quaternions.
```
Comment (hakonanes) [inline orix/quaternion/rotation.py:389]:
```suggestion
            The shape out of which to construct identity rotations.
```

File: orix/quaternion/rotation.py [hunk 6]
```diff
@@ -385,133 +380,26 @@ def dot_outer(self, other: Rotation) -> np.ndarray:
...
-            eu[..., 2] = np.where(not_zero, gamma, eu[..., 2])
-
-        # Reduce Euler angles to definition range
-        eu[np.abs(eu) < _FLOAT_EPS] = 0
-        eu = np.where(eu < 0, np.mod(eu + 2 * np.pi, (2 * np.pi, np.pi, 2 * np.pi)), eu)
-
-        return eu
-
-    # TODO: Remove decorator, **kwargs, and use of "convention" in 1.0
+    # TODO: Remove **kwargs, and use of "convention" in 1.0.
```
Comment (hakonanes) [inline orix/quaternion/rotation.py:400]:
```suggestion
    # TODO: Remove **kwargs in 1.0.
```

File: orix/quaternion/rotation.py [hunk 7]
```diff
@@ -526,167 +414,13 @@ def from_euler(
...
-            q[..., i] = np.where(q[..., 0] < 0, -q[..., i], q[..., i])
-
-        data = Quaternion(q)
-
-        if direction == "crystal2lab":
-            data = ~data
+        rot = super().from_euler(euler=euler, direction=direction, **kwargs)
 
-        rot = cls(data)
+        n = euler.shape[:-1]
```
Comment (hakonanes) [inline orix/quaternion/rotation.py:420]: I suggest to just do `rot.improper = np.zeros(euler.shape[:-1])`.

**Review by hakonanes — CHANGES_REQUESTED** [review]
Thank you for setting this up @anderscmathisen!

I've made just a few suggestions for improvements, mostly to docstring wordings I or other people originally added...

File: orix/quaternion/rotation.py (hunk 4 above)
Comment (anderscmathisen) [reply orix/quaternion/rotation.py:383]: Agreed, I suggest we also move this function to the Quaternion class.

File: orix/quaternion/quaternion.py [hunk 8]
```diff
@@ -498,3 +502,340 @@ def _outer_dask(
...
+        ----------
+        shape
+            The shape out of which to construct identity quaternions.
+
+        Returns
+        -------
+        rot
+            Identity quaternions.
+        """
+        data = np.zeros(shape + (4,))
```
Comment (hakonanes) [inline orix/quaternion/quaternion.py:839]: If `shape` is a tuple this won't work, have to do the check as in `random()` above.

**Review by hakonanes — CHANGES_REQUESTED** [review]
Looks good. As I can see it's just one check in `Quaternion.identity()` missing, and we can merge this!

**Comment by hakonanes** [comment]
We haven't considered that quaternions in orix are not by default unit quaternions, however rotations are unit quaternions. We ensure this by normalizing rotations upon initialization, while quaternions are not normalized.

The conversions we have implemented are based on the work by Rowenhorst et al. (2015) https://doi.org/10.1088/0965-0393/23/8/083501, where they in the abstract state that the transformations in that work including quaternions concern *unit* quaternions. Hence, I suggest that all new
1. `from_x()` methods return unit quaternions
2. `to_x()` methods normalize before returning `x`

I've already done this in my fork of your branch here, @anderscmathisen, among other changes like moving some rotation tests upstream to the quaternion test file. I suggest I make a PR to your branch, merge that, and then someone else has to review this PR since it includes my work...

**Comment by anderscmathisen** [comment]
Yeah, I agree with your suggestions and nice that you have implemented them in a branch. 
Please let me know if I need to do anything.

**Comment by hakonanes** [comment]
I can fix the conflits.

**Comment by hakonanes** [comment]
Don't know if you've fixed lots of git conflicts @anderscmathisen, but what I did was basically (on the `move_to_from_methods` branch):

```bash
> git pull anderscmathisen move_to_from_methods  # Update my local fork of your branch
> git merge develop
Auto-merging CHANGELOG.rst
CONFLICT (content): Merge conflict in CHANGELOG.rst
Auto-merging orix/quaternion/quaternion.py
Auto-merging orix/quaternion/rotation.py
CONFLICT (content): Merge conflict in orix/quaternion/rotation.py
Auto-merging orix/tests/quaternion/test_rotation.py
CONFLICT (content): Merge conflict in orix/tests/quaternion/test_rotation.py
Automatic merge failed; fix conflicts and then commit the result.
```

Open the conflicting files and do the necessary changes, in this case remove the changes on `develop` and keep the ones on `move_to_from_methods`, then:

```bash
> git add .
> git commit -s
> git push anderscmathisen move_to_from_methods  # Update your branch
```

**Comment by hakonanes** [comment]
I can add a small test covering the single line missing (https://coveralls.io/builds/53787279/source?filename=orix%2Fquaternion%2Frotation.py#L581).

**Comment by hakonanes** [comment]
Fixed another conflict.

**Review by pc494 — APPROVED** [review]
Sorry for the delay all, I'm happy with this. Because it's been a little while since the request, I'm going to let @hakonanes merge just in case something else has happened since that I'm not aware of.

**Pull request merged**